internal/recover/recover.go

```diff
@@ -0,0 +1,339 @@
+// go-dqlite's info.yaml is excluded from the tarball.
+func CreateRecoveryTarball(filesystem *sys.OS) error {
+	dbFS := os.DirFS(filesystem.DatabaseDir)
+	dbFiles, err := fs.Glob(dbFS, "*")
+	if err != nil {
+		return fmt.Errorf("%w", err)
+	}
+
+	tarballPath := path.Join(filesystem.StateDir, "recovery_db.tar.gz")
+
+	// info.yaml is used by go-dqlite to keep track of the current cluster member's
+	// ID and address. We shouldn't replicate the recovery member's info.yaml
+	// to all other members, so exclude it from the tarball:
+	for indx, filename := range dbFiles {
+		if filename == "info.yaml" {
+			newlen := len(dbFiles) - 1
+			dbFiles[indx] = dbFiles[newlen]
+			dbFiles = dbFiles[:newlen]
+			break
+		}
+	}
+
+	return createTarball(tarballPath, filesystem.DatabaseDir, dbFiles)
+}
+
+// MaybeUnpackRecoveryTarball checks for the presence of a recovery tarball in
+// fiesystem.StateDir. If it exists, unpack it into a temporary directory,
+// ensure that it is a valid microcluster recovery tarball, and replace the
+// existing filesystem.DatabaseDir.
+func MaybeUnpackRecoveryTarball(filesystem *sys.OS) error {
+	tarballPath := path.Join(filesystem.StateDir, "recovery_db.tar.gz")
+	unpackDir := path.Join(filesystem.StateDir, "recovery_db")
+
+	// Determine if the recovery tarball exists
+	if _, err := os.Stat(tarballPath); errors.Is(err, os.ErrNotExist) {
+		return nil
+	}
+
+	logger.Warn("Recovery tarball located; attempting DB recovery", logger.Ctx{"tarball": tarballPath})
+
+	err := unpackTarball(tarballPath, unpackDir)
+	if err != nil {
+		return err
+	}
+
+	// TODO Is this a reasonable sanity check or is a parseable cluster.yaml
+	// containing the local node sufficient?
+	metadataPath := path.Join(unpackDir, "metadata1")
+	if _, err := os.Stat(metadataPath); errors.Is(err, os.ErrNotExist) {
+		return fmt.Errorf("missing %q in recovery tarball", metadataPath)
+	}
+
+	// sanity check: valid cluster.yaml in the incoming DB dir
+	clusterYamlPath := path.Join(unpackDir, "cluster.yaml")
+	incomingNodeInfo, err := dumpYamlNodeStore(clusterYamlPath)
+	if err != nil {
+		return err
+	}
+
+	// use the local info.yaml so that the dqlite ID is preserved on each
+	// cluster member
+	localInfoYamlPath := path.Join(filesystem.DatabaseDir, "info.yaml")
+	recoveryInfoYamlPath := path.Join(unpackDir, "info.yaml")
+
+	localInfoYaml, err := os.ReadFile(localInfoYamlPath)
+	if err != nil {
+		return err
+	}
+
+	var localInfo dqlite.NodeInfo
+	err = yaml.Unmarshal(localInfoYaml, &localInfo)
+	if err != nil {
+		return fmt.Errorf("invalid %q", localInfoYamlPath)
+	}
+
+	found := false
+	for _, incomingInfo := range incomingNodeInfo {
+		found = localInfo.ID == incomingInfo.ID
+
+		if found {
+			break
+		}
+	}
+
+	if !found {
+		return fmt.Errorf("missing local cluster member in incoming cluster.yaml")
+	}
+
+	err = os.WriteFile(recoveryInfoYamlPath, localInfoYaml, 0o664)
+	if err != nil {
+		return err
+	}
+
+	// FIXME: Take a DB backup
```

Comment:
I think this should definitely get implemented before this PR is merged.

Reply:
Will add the commits from #160 here.